Mage.Sets/src/mage/cards/k/KjeldoranEliteGuard.java

```diff
@@ -0,0 +1,128 @@
+                Zone.BATTLEFIELD,
+                new KjeldoranEliteGuardEffect(),
+                new TapSourceCost(),
+                new IsPhaseCondition(TurnPhase.COMBAT, false));
+        ability.addTarget(new TargetControlledCreaturePermanent());
+        this.addAbility(ability);
+    }
+
+    private KjeldoranEliteGuard(final KjeldoranEliteGuard card) { super(card); }
+
+    @Override
+    public Card copy() {
+        return new KjeldoranEliteGuard(this);
+    }
+}
+
+class KjeldoranEliteGuardEffect extends OneShotEffect {
+
+    KjeldoranEliteGuardEffect() {
+        super(Outcome.Neutral);
+        staticText = "Target creature gets +2/+2 until end of turn. "
+                + "When that creature leaves the battlefield this turn, sacrifice Kjeldoran Elite Guard. "
+                + "Activate only during combat.";
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        if (source == null) { return false; }
+
+        Permanent targetCreature = game.getPermanent(source.getFirstTarget());
+        if (targetCreature == null) { return false; }
+
+        // Target creature gets +2/+2 until end of turn.
+        BoostTargetEffect buffEffect = new BoostTargetEffect(2, 2, Duration.EndOfTurn);
+        buffEffect.setTargetPointer(new FixedTarget(source.getFirstTarget(), game));
+        game.addEffect(buffEffect, source);
+
+        // When that creature leaves the battlefield this turn, sacrifice Kjeldoran Elite Guard.
+        SacrificeTargetEffect sacrificeKjeldoranEliteGuardEffect = new SacrificeTargetEffect();
+        sacrificeKjeldoranEliteGuardEffect.setTargetPointer(new FixedTarget(source.getSourceId(), game));
+        game.addDelayedTriggeredAbility(
+                new KjeldoranEliteGuardDelayedTriggeredAbility(
+                        sacrificeKjeldoranEliteGuardEffect,
+                        source.getFirstTarget()),
+                source);
+
+        return true;
+    }
+
+    private KjeldoranEliteGuardEffect(KjeldoranEliteGuardEffect effect) { super(effect); }
+
+    @Override
+    public KjeldoranEliteGuardEffect copy() { return new KjeldoranEliteGuardEffect(this); }
+}
+
+class KjeldoranEliteGuardDelayedTriggeredAbility extends DelayedTriggeredAbility {
+
+    UUID creatureId;
+
+    KjeldoranEliteGuardDelayedTriggeredAbility(Effect effect, UUID creatureId) {
+        super(effect, Duration.EndOfTurn, true);
+        this.creatureId = creatureId;
+    }
+
+    KjeldoranEliteGuardDelayedTriggeredAbility(KjeldoranEliteGuardDelayedTriggeredAbility ability) {
+        super(ability);
+        this.creatureId = ability.creatureId;
+    }
+
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE
```

Comment:
you don't need to check event type here, it's only ever gonna be `ZONE_CHANGE` if you get to this point